### PR TITLE
Use default values when switching anyOf option

### LINKED ISSUE #4367: Unable to switch between the options in the anyOf when omitExtraData and liveOmit is true
### Prerequisites

- [X] I have searched the existing issues
- [X] I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [X] I have read the [documentation](https://rjsf-team.github.io/react-jsonschema-form/docs)
- [X] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net), [Codesandbox.io](https://codesandbox.io) or preferably a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### What theme are you using?

core

### Version

5.x

### Current Behavior

I have a schema that has anyOf between two options. The form renders a selector for it, but when I use the selector to select the options, it does not work. I also notice when I select the option 2 for example, the formData is filled with default of both option 1 and option 2.

When turning off omitExtraData and liveOmit, it works normally.

### Expected Behavior

When I switch to option 2, the fields for option 2 should be rendered, and the formData contains only the default values for option 2

### Steps To Reproduce

Visit [playground](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJteVByb3BlcnR5Ijp7InZhbHVlX3Byb3ZpZGVyX3R5cGUiOiJUWVBFXzEiLCJzdGFydCI6NjUwMDAwMDAwMCwic3RvcCI6ODUwMDAwMDAwMCwic3RlcF9zaXplIjoxMDAwMDAsIm51bV9zYW1wbGVzIjoyMDAxLCJzd2VlcF9zaGlmdCI6MH19LCJzY2hlbWEiOnsidGl0bGUiOiJTZXR0aW5ncyIsInR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7Im15UHJvcGVydHkiOnsiZGVmYXVsdCI6eyJ2YWx1ZV9wcm92aWRlcl90eXBlIjoiVFlQRV8xIiwic3RhcnQiOjY1MDAwMDAwMDAsInN0b3AiOjg1MDAwMDAwMDAsInN0ZXBfc2l6ZSI6MTAwMDAwLCJudW1fc2FtcGxlcyI6MjAwMX0sImFueU9mIjpbeyIkcmVmIjoiIy9kZWZpbml0aW9ucy9UeXBlMiJ9LHsiJHJlZiI6IiMvZGVmaW5pdGlvbnMvVHlwZTEifV19fSwiZGVmaW5pdGlvbnMiOnsiVHlwZTIiOnsidGl0bGUiOiJUeXBlMiIsInR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7InN3ZWVwX3JhbmdlIjp7InRpdGxlIjoiU3dlZXAgcmFuZ2UiLCJ0eXBlIjoibnVtYmVyIn0sInN3ZWVwX3N0ZXBfc2l6ZSI6eyJ0aXRsZSI6IlN0ZXAgc2l6ZSIsInR5cGUiOiJudW1iZXIifSwic3dlZXBfc2hpZnQiOnsidGl0bGUiOiJTaGlmdCIsImRlZmF1bHQiOjAsInR5cGUiOiJudW1iZXIifX0sInJlcXVpcmVkIjpbInN3ZWVwX3JhbmdlIiwic3dlZXBfc3RlcF9zaXplIl19LCJUeXBlMSI6eyJ0aXRsZSI6IlR5cGUxIiwidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsic3RhcnQiOnsidGl0bGUiOiJTdGFydCIsInR5cGUiOiJudW1iZXIifSwic3RvcCI6eyJ0aXRsZSI6IlN0b3AiLCJ0eXBlIjoibnVtYmVyIn0sInN0ZXBfc2l6ZSI6eyJ0aXRsZSI6IlN0ZXAgc2l6ZSIsInR5cGUiOiJudW1iZXIifSwibnVtX3NhbXBsZXMiOnsidGl0bGUiOiJOdW1iZXIgb2Ygc2FtcGxlcyIsInR5cGUiOiJpbnRlZ2VyIn0sInZhbHVlX3Byb3ZpZGVyX3R5cGUiOnsidGl0bGUiOiJWYWx1ZSBQcm92aWRlciBUeXBlIiwiZGVmYXVsdCI6IlRZUEVfMSIsImVudW0iOlsiVFlQRV8xIl0sInR5cGUiOiJzdHJpbmcifX0sInJlcXVpcmVkIjpbInN0YXJ0Iiwic3RvcCIsInZhbHVlX3Byb3ZpZGVyX3R5cGUiXX19fSwidWlTY2hlbWEiOnt9LCJ0aGVtZSI6ImRlZmF1bHQiLCJsaXZlU2V0dGluZ3MiOnsic2hvd0Vycm9yTGlzdCI6InRvcCIsInZhbGlkYXRlIjpmYWxzZSwiZGlzYWJsZWQiOmZhbHNlLCJub0h0bWw1VmFsaWRhdGUiOmZhbHNlLCJyZWFkb25seSI6ZmFsc2UsIm9taXRFeHRyYURhdGEiOnRydWUsImxpdmVPbWl0Ijp0cnVlLCJleHBlcmltZW50YWxfZGVmYXVsdEZvcm1TdGF0ZUJlaGF2aW9yIjp7ImFycmF5TWluSXRlbXMiOiJwb3B1bGF0ZSIsImFsbE9mIjoic2tpcERlZmF1bHRzIiwiZW1wdHlPYmplY3RGaWVsZHMiOiJwb3B1bGF0ZUFsbERlZmF1bHRzIiwibWVyZ2VEZWZhdWx0c0ludG9Gb3JtRGF0YSI6InVzZUZvcm1EYXRhSWZQcmVzZW50In0sImxpdmVWYWxpZGF0ZSI6ZmFsc2UsIm5vVmFsaWRhdGUiOmZhbHNlfX0=), and try to use the selector to select option 2. You will notice the fields for option 1 still render and the formData includes the default values for both option 1 and option 2

### Environment

```markdown
Playground: https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJteVByb3BlcnR5Ijp7InZhbHVlX3Byb3ZpZGVyX3R5cGUiOiJUWVBFXzEiLCJzdGFydCI6NjUwMDAwMDAwMCwic3RvcCI6ODUwMDAwMDAwMCwic3RlcF9zaXplIjoxMDAwMDAsIm51bV9zYW1wbGVzIjoyMDAxLCJzd2VlcF9zaGlmdCI6MH19LCJzY2hlbWEiOnsidGl0bGUiOiJTZXR0aW5ncyIsInR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7Im15UHJvcGVydHkiOnsiZGVmYXVsdCI6eyJ2YWx1ZV9wcm92aWRlcl90eXBlIjoiVFlQRV8xIiwic3RhcnQiOjY1MDAwMDAwMDAsInN0b3AiOjg1MDAwMDAwMDAsInN0ZXBfc2l6ZSI6MTAwMDAwLCJudW1fc2FtcGxlcyI6MjAwMX0sImFueU9mIjpbeyIkcmVmIjoiIy9kZWZpbml0aW9ucy9UeXBlMiJ9LHsiJHJlZiI6IiMvZGVmaW5pdGlvbnMvVHlwZTEifV19fSwiZGVmaW5pdGlvbnMiOnsiVHlwZTIiOnsidGl0bGUiOiJUeXBlMiIsInR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7InN3ZWVwX3JhbmdlIjp7InRpdGxlIjoiU3dlZXAgcmFuZ2UiLCJ0eXBlIjoibnVtYmVyIn0sInN3ZWVwX3N0ZXBfc2l6ZSI6eyJ0aXRsZSI6IlN0ZXAgc2l6ZSIsInR5cGUiOiJudW1iZXIifSwic3dlZXBfc2hpZnQiOnsidGl0bGUiOiJTaGlmdCIsImRlZmF1bHQiOjAsInR5cGUiOiJudW1iZXIifX0sInJlcXVpcmVkIjpbInN3ZWVwX3JhbmdlIiwic3dlZXBfc3RlcF9zaXplIl19LCJUeXBlMSI6eyJ0aXRsZSI6IlR5cGUxIiwidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsic3RhcnQiOnsidGl0bGUiOiJTdGFydCIsInR5cGUiOiJudW1iZXIifSwic3RvcCI6eyJ0aXRsZSI6IlN0b3AiLCJ0eXBlIjoibnVtYmVyIn0sInN0ZXBfc2l6ZSI6eyJ0aXRsZSI6IlN0ZXAgc2l6ZSIsInR5cGUiOiJudW1iZXIifSwibnVtX3NhbXBsZXMiOnsidGl0bGUiOiJOdW1iZXIgb2Ygc2FtcGxlcyIsInR5cGUiOiJpbnRlZ2VyIn0sInZhbHVlX3Byb3ZpZGVyX3R5cGUiOnsidGl0bGUiOiJWYWx1ZSBQcm92aWRlciBUeXBlIiwiZGVmYXVsdCI6IlRZUEVfMSIsImVudW0iOlsiVFlQRV8xIl0sInR5cGUiOiJzdHJpbmcifX0sInJlcXVpcmVkIjpbInN0YXJ0Iiwic3RvcCIsInZhbHVlX3Byb3ZpZGVyX3R5cGUiXX19fSwidWlTY2hlbWEiOnt9LCJ0aGVtZSI6ImRlZmF1bHQiLCJsaXZlU2V0dGluZ3MiOnsic2hvd0Vycm9yTGlzdCI6InRvcCIsInZhbGlkYXRlIjpmYWxzZSwiZGlzYWJsZWQiOmZhbHNlLCJub0h0bWw1VmFsaWRhdGUiOmZhbHNlLCJyZWFkb25seSI6ZmFsc2UsIm9taXRFeHRyYURhdGEiOnRydWUsImxpdmVPbWl0Ijp0cnVlLCJleHBlcmltZW50YWxfZGVmYXVsdEZvcm1TdGF0ZUJlaGF2aW9yIjp7ImFycmF5TWluSXRlbXMiOiJwb3B1bGF0ZSIsImFsbE9mIjoic2tpcERlZmF1bHRzIiwiZW1wdHlPYmplY3RGaWVsZHMiOiJwb3B1bGF0ZUFsbERlZmF1bHRzIiwibWVyZ2VEZWZhdWx0c0ludG9Gb3JtRGF0YSI6InVzZUZvcm1EYXRhSWZQcmVzZW50In0sImxpdmVWYWxpZGF0ZSI6ZmFsc2UsIm5vVmFsaWRhdGUiOmZhbHNlfX0=
```

### Anything else?

The schema I'm testing with

```
{
  "title": "Settings",
  "type": "object",
  "properties": {
    "myProperty": {
      "default": {
        "value_provider_type": "TYPE_1",
        "start": 6500000000,
        "stop": 8500000000,
        "step_size": 100000,
        "num_samples": 2001
      },
      "anyOf": [
        {
          "$ref": "#/definitions/Type2"
        },
        {
          "$ref": "#/definitions/Type1"
        }
      ]
    }
  },
  "definitions": {
    "Type2": {
      "title": "Type2",
      "type": "object",
      "properties": {
        "sweep_range": {
          "title": "Sweep range",
          "type": "number"
        },
        "sweep_step_size": {
          "title": "Step size",
          "type": "number"
        },
        "sweep_shift": {
          "title": "Shift",
          "default": 0,
          "type": "number"
        }
      },
      "required": [
        "sweep_range",
        "sweep_step_size"
      ]
    },
    "Type1": {
      "title": "Type1",
      "type": "object",
      "properties": {
        "start": {
          "title": "Start",
          "type": "number"
        },
        "stop": {
          "title": "Stop",
          "type": "number"
        },
        "step_size": {
          "title": "Step size",
          "type": "number"
        },
        "num_samples": {
          "title": "Number of samples",
          "type": "integer",
          "default": -1
        },
        "value_provider_type": {
          "title": "Value Provider Type",
          "default": "TYPE_1",
          "enum": [
            "TYPE_1"
          ],
          "type": "string"
        }
      },
      "required": [
        "start",
        "stop",
        "value_provider_type"
      ]
    }
  }
}
```

### PR DESCRIPTION
**Pull request by quentin-sommer**
### Reasons for making this change

`sanitizeDataForNewSchema` is called when switching the option selected from a anyOf field. It doesn't check default values in the new schema. This leads to the formData being cleared completely. This PR makes it respect default values.

Maybe this is a bad idea. The implementation seemed very aware that it wasn't handling cases for scalar values. Let me know what you think about it.

The UI wasn't properly handling the option change resulting in undefined data (See my video in #4367).

In summary this change makes this schema work as expected:
```json
{
  "additionalProperties": false,
  "properties": {
    "any_of_array_or_null": {
      "anyOf": [
        {
          "items": {
            "type": "integer"
          },
          "type": "array",
          "title": "Array value selected"
        },
        {
          "type": "null",
          "title": "Null value selected"
        }
      ]
    }
  },
  "required": [
    "any_of_array_or_null"
  ],
  "type": "object"
}
```

fixes #4367

### Checklist

- [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://rjsf-team.github.io/react-jsonschema-form/docs/contributing) of the Markdown text I've added
- [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests. I've run `npx nx run-many --target=build --exclude=@rjsf/docs && npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR
- [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

File: packages/utils/src/schema/sanitizeDataForNewSchema.ts [hunk 1]
```diff
@@ -193,5 +196,12 @@ export default function sanitizeDataForNewSchema<
     }
     // Also probably want to deal with `prefixItems` as tuples with the latest 2020 draft
   }
+  // Schema contains a single scalar value
+  else {
```
Comment (quentin-sommer) [inline packages/utils/src/schema/sanitizeDataForNewSchema.ts:200]: I'm not sure what are the other possible cases here. 

Logic before this branch:
- We checked that the schema  doesn't have a `properties` key
- This check is false old `schema type == array AND new schema type == array AND current data == array`

I'm only checking that newSchema has a `default` key but maybe there is a better check here?

**Comment by nickgros** [comment]
@quentin-sommer A couple of thoughts:

Does your change fix the original issue in #4367 ? When I tried it locally it did not seem to fix it, but I want to confirm with you.

Instead of changing `sanitizeDataForNewSchema`, it seems like we could let `getDefaultFormState` handle this case. Since `sanitizeDataForNewSchema` is only called in `MultiSchemaField`, what do you think of changing line MultiSchemaField L126 from

```ts
    if (newFormData && newOption) {
```
to
```ts
    if (newOption) {
```

Your test seems to pass with this change, but I want to make sure I'm not missing an edge case I am unaware of.

**Comment by quentin-sommer** [comment]
I applied your suggestion and it fixes the original issue in #4367 as well as the schema in this PR description, thanks!

I added a changelog entry

**Review by heath-freenome — APPROVED** [review]

File: packages/core/test/anyOf.test.jsx [hunk 2]
```diff
@@ -165,6 +165,43 @@ describe('anyOf', () => {
     );
   });
 
+  it('should assign a default value and set defaults on option change for scalar types schemas', () => {
```
Comment (heath-freenome) [inline packages/core/test/anyOf.test.jsx:168]: Can you add a similar test for the `oneOf.test.jsx` file?

Comment (quentin-sommer) [reply packages/core/test/anyOf.test.jsx:168]: Sure thing, PR update 👍

**Review by heath-freenome — APPROVED** [review]

**Comment by quentin-sommer** [comment]
@heath-freenome @nickgros Are there next steps before we can merge this?

**Comment by heath-freenome** [comment]
Sorry, the holidays got in the way of this getting merged

**Comment by Alan-Cha** [comment]
@quentin-sommer Your fix seems to address issues with `anyOf` with arrays and objects but it doesn't seem to work with primitives. Try the following schema:

```json
{
  "type": "object",
  "properties": {
    "working_array": {
      "anyOf": [
        {
          "type": "array",
          "items": {
            "type": "integer"
          },
          "title": "array"
        },
        {
          "type": "null",
          "title": "null"
        }
      ]
    },
    "working_object": {
      "anyOf": [
        {
          "type": "object",
          "title": "object"
        },
        {
          "type": "null",
          "title": "null"
        }
      ]
    },
    "not_working_string": {
      "anyOf": [
        {
          "type": "string",
          "title": "string"
        },
        {
          "type": "null",
          "title": "null"
        }
      ]
    },
    "not_working_integer": {
      "anyOf": [
        {
          "type": "integer",
          "title": "integer"
        },
        {
          "type": "null",
          "title": "null"
        }
      ]
    }
  }
}
```

https://github.com/user-attachments/assets/eda2955c-acf9-457b-80e8-afebc7846b9a

As you can see, the array and object seems to be working as expected but not for strings and integers.

With the array and object, we can switch between the respective value and `null`.

With string, may be initially possible to add string data but after switching to `null`, we can no longer add string data any more.

With integer, it seems to be stuck providing `null`. It doesn't seem to be possible to change its value to an integer.

I am assuming that this is due to primitives having different default value behavior than arrays and objects. Do you have any thoughts?

**Comment by quentin-sommer** [comment]
I also see the issue. My PR addressed the cases where we have a default value provided in the schema. Unfortunately I don't have enough knowledge in the library to know what happens when there isn't a default value.

For example I have no idea why the `not_working_integer` is initialized to a `null` in formData but the `not_working_string` is not initialized at all at the start.
